src/gluonts/itertools.py

```diff
@@ -140,12 +145,150 @@ class Chain:
+    """
+
+    item: int
+    local: int
+
+
+@dataclass
+class Fuse:
+    """Fuse collections together to act as single collections.
+
+    >>> a = [0, 1, 2]
+    >>> b = [3, 4, 5]
+    >>> fused = Fuse([a, b])
+    >>> assert len(a) + len(b) == len(fused)
+    >>> list(fused[2:5])
+    [2, 3, 4]
+    >>> list(fused[3:])
+    [3, 4, 5]
+
+    This is similar to ``Chain``, but also allows to select directly into the
+    data. While ``Chain`` creates a single ``Iterable`` out of a set of
+    ``Iterable``s, ``Fuse`` creates a single ``Collection`` out of a set of
+    ``Collection``s.
+    """
+
+    collections: List[Collection]
+    _lengths: List[int] = field(default_factory=list)
+
+    def __post_init__(self):
+        if not self._lengths:
+            self._lengths = list(map(len, self.collections))
+
+        self._length = sum(self._lengths)
+        self._offsets = np.cumsum(self._lengths)
+
+    def __len__(self):
+        return self._length
+
+    def _get_range(self, start: _SubIndex, stop: _SubIndex) -> "Fuse":
+        first = self.collections[start.item]
+
+        if start.item == stop.item:
+            return Fuse([first[start.local : stop.local]])
+
+        items = []
+
+        first = first[start.local :]
+        if len(first) > 0:
+            items.append(first)
+
+        for item_index in range(start.item + 1, stop.item):
+            items.append(self.collections[item_index])
+
+        items.append(self.collections[stop.item][: stop.local])
+        return Fuse(items)
+
+    def _location_for(self, idx, side="right") -> _SubIndex:
+        """Map global index to pair of index to collection and index within
+        that collection.
+
+        >>> fuse = Fuse([[0, 0], [1, 1]])
+        >>> fuse._location_for(0)
+        _SubIndex(item=0, local=0)
+        >>> fuse._location_for(1)
+        _SubIndex(item=0, local=1)
+        >>> fuse._location_for(2)
+        _SubIndex(item=1, local=0)
+        >>> fuse._location_for(3)
+        _SubIndex(item=1, local=1)
+
+        """
+        if idx == 0:
+            return _SubIndex(0, 0)
+
+        part_no = np.searchsorted(self._offsets, idx, side)
+
+        if part_no == 0:
+            local_idx = idx
+        elif part_no >= len(self.collections):
+            part_no -= 1
+            local_idx = len(self.collections[part_no])
+        else:
+            local_idx = idx - self._offsets[part_no - 1]
+
+        return _SubIndex(part_no, local_idx)
+
+    def __getitem__(self, idx):
+        if isinstance(idx, slice):
+            start, stop, step = idx.indices(len(self))
+            assert step is None or step == 1
+
+            start_index = self._location_for(start)
+            stop_index = self._location_for(stop)
+            return self._get_range(start_index, stop_index)
+
+        subindex = self._location_for(idx)
+        return self.collections[subindex.item][subindex.local]
+
+    def split(self, indices: List[int]) -> List[Fuse]:
```

Comment:
This could be defined for generic collections maybe?

Reply:
What do you mean?